Teleport session improvements

kontsevoy opened the pull request:
Two changes:
- e349dcc - Make sure all active sessions are properly closed when a `server.Close()` is called.
- 8009984 - Replay last chunk to a new party who joins an existing session.

kontsevoy committed:
Cosmetic code refactoring + clearer log messges

kontsevoy committed:
Added a more robust session cleanup

When someone calls sshserver.Close() it will properly close
all active sessions.

kontsevoy committed:
Last chunk replay on session.join()

When a new user (party) joins an existing SSH session, Teleport will
replay the last chunk into party's io.writer

This is to improve user's experience, so you'll see what's going on upon
joining.

klizhentas commented on lib/srv/sess.go:
```diff
@@ -419,11 +426,39 @@ func (s *session) addParty(p *party) {
+
+	// when a new user joins the session, it's visually helpful to replay the last piece of output so they
+	// will see the same prompt the host is seeing
+	replayLastChunk := func() {
```
can you move this function to be a method? seems a bit generic to be a closure

kontsevoy replied:
mm... actually I think this must be deleted alltogether. to be discussed.

klizhentas commented:
lgtm with comment about a closure/method

kontsevoy committed:
Changed how "most recent chunk" is replayed

Every session now keeps a "session buffer", which implements `io.Reader`
and `recorder.ChunkWriter` interfaces.

This allows:

- Replay the most recent chunk to joinig parties
- Stream read-only sessions (not implemented yet)

This is better than my previous implementation because:

- It doesn't hit auth servers: everything is local and in-memory
- It doesn't require additional privileges for role.Node. I wasn't
  comfortable for all nodes having full access to all past sessions.

kontsevoy commented:
I changed the implementation. See the comments for f81bfdb

klizhentas commented:
lgtm
